UI/window-basic-preview.cpp

```diff
@@ -1091,6 +1136,170 @@ static void DrawCircleAtPos(float x, float y, matrix4 &matrix,
+
+	vec3 labelSize, labelPos;
+	vec3_set(&labelSize,
+		obs_source_get_width(sizeLabel),
+		obs_source_get_height(sizeLabel),
+		1.0f);
+	vec3_div(&labelSize, &labelSize, &viewport);
+
+	float minSize = 0.0f;
+	vec3_set(&labelPos, end.x, end.y, end.z);
+	if ((end.x - start.x) >= (end.y - start.y)) {
+		// Horizontal line
+		labelPos.x -= (end.x - start.x) / 2;
+		labelPos.x -= labelSize.x / 2;
+		labelPos.y += labelMargin;
+		minSize = labelSize.x * virtualLabelSizeFactor;
+	} else {
+		// Vertical line
+		labelPos.y -= (end.y - start.y) / 2;
+		labelPos.y -= labelSize.y / 2;
+		labelPos.x += labelMargin;
+		minSize = labelSize.y * virtualLabelSizeFactor;
+	}
+
+	if (abs(length) >= minSize) {
+		DrawLabel(labelPos, sizeLabel, viewport);
+	}
+
+	currentSizeLabel++;
+	if (currentSizeLabel > (PREVIEW_SPACING_LABEL_COUNT - 1)) {
+		currentSizeLabel = 0;
+	}
+}
+
+static void boxCoordsToView(vec3 *coord, matrix4 &boxTransform,
+	vec3 &viewport, float previewScale)
+{
+	vec3_transform(coord, coord, &boxTransform);
+	vec3_div(coord, coord, &viewport);
+	vec3_mulf(coord, coord, previewScale);
+}
+
+void OBSBasicPreview::DrawSpacingHelpers(matrix4 &boxTransform, vec3 &viewport,
+	float previewScale)
+{
+	gs_matrix_push();
+	gs_matrix_identity();
+	gs_matrix_scale(&viewport);
+
+	// Compute positions of TRBL points of item box
+	vec3 boxLeft, boxRight, boxTop, boxBottom;
+	vec3_set(&boxLeft, 0.0f, 0.5f, 1.0f);
+	vec3_set(&boxRight, 1.0f, 0.5f, 1.0f);
+	vec3_set(&boxTop, 0.5f, 0.0f, 1.0f);
+	vec3_set(&boxBottom, 0.5f, 1.0f, 1.0f);
+
+	boxCoordsToView(&boxLeft, boxTransform, viewport, previewScale);
+	boxCoordsToView(&boxRight, boxTransform, viewport, previewScale);
+	boxCoordsToView(&boxTop, boxTransform, viewport, previewScale);
+	boxCoordsToView(&boxBottom, boxTransform, viewport, previewScale);
+
+	float boxWidth =
+		sqrt(pow(boxLeft.x - boxRight.x, 2) + pow(boxLeft.y - boxRight.y, 2));
+	float boxHeight =
+		sqrt(pow(boxTop.x - boxBottom.x, 2) + pow(boxTop.y - boxBottom.y, 2));
+	if (boxWidth > 0.0f || boxHeight > 0.0f) {
+		// View borders
+		vec3 viewLeft, viewRight, viewTop, viewBottom;
+		vec3_set(&viewLeft, 0.0f, boxLeft.y, 1.0f);
+		vec3_set(&viewRight, 1.0f, boxRight.y, 1.0f);
+		vec3_set(&viewTop, boxTop.x, 0.0f, 1.0f);
+		vec3_set(&viewBottom, boxBottom.x, 1.0f, 1.0f);
+
+		// Clip lines to view boundaries
+		vec3_max(&boxLeft, &boxLeft, &viewLeft);
+		vec3_min(&boxRight, &boxRight, &viewRight);
+		vec3_max(&boxTop, &boxTop, &viewTop);
+		vec3_min(&boxBottom, &boxBottom, &viewBottom);
+
+		// Draw helper lines
+		vec4 green;
+		vec4_set(&green, 0.0f, 1.0f, 0.0f, 1.0f);
+
+		gs_effect_t *eff = gs_get_effect();
+		gs_eparam_t *eparam = gs_effect_get_param_by_name(eff, "color");
+		gs_effect_set_vec4(eparam, &green);
+
+		DrawLine(viewTop, boxTop);
+		DrawLine(viewRight, boxRight);
+		DrawLine(viewBottom, boxBottom);
+		DrawLine(viewLeft, boxLeft);
+
+		DrawSingleSpacingHelper(viewTop, boxTop, viewport);
```

Comment:
I'd like to put DrawLine in DrawSingleSpacingHelper. But when I do so, the line is hidden while the label is visible. Did I miss something?